File: lab2/sierpinski.py
from OpenGL.GL import *

import utils

dimensions = utils.Coord(1000, 1000)
bounds = utils.Coord(100, 100)
viewport = utils.Viewport(bounds)

background = [0.9, 0.9, 0.9]


def startup():
    glClearColor(background[0], background[1], background[2], 1)
    viewport(None, dimensions)


def shutdown():
    pass


def make_sierpinski_carpet(center: utils.Coord, sides: utils.Coord, level):
    if level <= 0:
        return

    new_sides = utils.Coord(sides.x / 3, sides.y / 3)

    utils.make_rectangle(center, sides, [0.5, 0.3, 0.6])
    utils.make_rectangle(center, new_sides, background)

    make_sierpinski_carpet(utils.Coord(center.x - new_sides.x, center.y + new_sides.y), new_sides, level - 1)
    make_sierpinski_carpet(utils.Coord(center.x, center.y + new_sides.y), new_sides, level - 1)
    make_sierpinski_carpet(utils.Coord(center.x + new_sides.x, center.y + new_sides.y), new_sides, level - 1)
    make_sierpinski_carpet(utils.Coord(center.x + new_sides.x, center.y), new_sides, level - 1)
    make_sierpinski_carpet(utils.Coord(center.x + new_sides.x, center.y - new_sides.y), new_sides, level - 1)
    make_sierpinski_carpet(utils.Coord(center.x, center.y - new_sides.y), new_sides, level - 1)
    make_sierpinski_carpet(utils.Coord(center.x - new_sides.x, center.y - new_sides.y), new_sides, level - 1)
    make_sierpinski_carpet(utils.Coord(center.x - new_sides.x, center.y), new_sides, level - 1)


def render(time):
    glClear(GL_COLOR_BUFFER_BIT)
    make_sierpinski_carpet(utils.Coord(0, 0), utils.Coord(200, 200), 5)
    glFlush()


def main():
    utils.default_init(dimensions=dimensions,
                       startup=startup,
                       render=render,
                       shutdown=shutdown,
                       viewport=viewport)


if __name__ == '__main__':
    main()

File: lab2/Sierpinski.cpp
#if defined(_WIN32)
#if !defined(NOMINMAX)
#define NOMINMAX
#endif
#include <Windows.h>
#endif
#include <GLFW/glfw3.h>

#include <array>

#include "Sierpinski.h"

namespace gk::lab2
{

auto Sierpinski::run(Vector2i dimensions, uint32_t level, bool vsync) -> void
{
    this->dimensions = dimensions;
    this->level = level;

    utils::configureViewport(bounds, nearDepth, farDepth);
    utils::defaultInit(dimensions,
                       [this] { startup(); },
                       [this](auto time) { render(time); },
                       [this] { shutdown(); },
                       utils::Config::Dimensions::_2D,
                       vsync);
}

auto Sierpinski::startup() const -> void
{
    glClearColor(0.5f, 0.5f, 0.5f, 0.1f);
    utils::viewport(nullptr, dimensions.x, dimensions.y);
}

auto Sierpinski::shutdown() -> void
{

}

auto Sierpinski::render([[maybe_unused]] double time) -> void
{
    glClear(GL_COLOR_BUFFER_BIT);
    makeSierpinskiCarpet({0.f, 0.f}, {200.f, 200.f}, level);
    glFlush();
}

auto Sierpinski::makeSierpinskiCarpet(Vector2f center, Vector2f sides, uint32_t currentLevel) -> void
{
    if (currentLevel == 0)
    {
        return;
    }

    const auto newSides = Vector2f {sides.x / 3.f, sides.y / 3.f};

    makeRectangle(center, sides, {128, 40, 200});
    makeRectangle(center, newSides, {128, 128, 128});

    makeSierpinskiCarpet(Vector2f{center.x - newSides.x, center.y + newSides.y}, newSides, currentLevel - 1);
    makeSierpinskiCarpet(Vector2f{center.x, center.y + newSides.y}, newSides, currentLevel - 1);
    makeSierpinskiCarpet(Vector2f{center.x + newSides.x, center.y + newSides.y}, newSides, currentLevel - 1);
    makeSierpinskiCarpet(Vector2f{center.x + newSides.x, center.y}, newSides, currentLevel - 1);
    makeSierpinskiCarpet(Vector2f{center.x + newSides.x, center.y - newSides.y}, newSides, currentLevel - 1);
    makeSierpinskiCarpet(Vector2f{center.x, center.y - newSides.y}, newSides, currentLevel - 1);
    makeSierpinskiCarpet(Vector2f{center.x - newSides.x, center.y - newSides.y}, newSides, currentLevel - 1);
    makeSierpinskiCarpet(Vector2f{center.x - newSides.x, center.y}, newSides, currentLevel - 1);
}

auto Sierpinski::makeRectangle(Vector2f center, Vector2f sides, std::array<uint8_t, 3> color) -> void
{
    glBegin(GL_TRIANGLE_FAN);

    glColor3ubv(color.data());

    glVertex2f(center.x - (sides.x / 2), center.y + (sides.y / 2));
    glVertex2f(center.x + (sides.x / 2), center.y + (sides.y / 2));
    glVertex2f(center.x + (sides.x / 2), center.y - (sides.y / 2));
    glVertex2f(center.x - (sides.x / 2), center.y - (sides.y / 2));

    glEnd();
}

}